### DISCUSSION TITLE
Use GHA with a custom runner to run automated tests

### DISCUSSION THREAD

**Discussion by Josverl** [pycom/pymakr-vsc#228] [Ideas]
to improve quality ive started to set-up a simple hardware CI test 
- Pi4 
  - configured with [DietPi](https://dietpi.com/)
  - installed [GHA self hosted runner](https://docs.github.com/en/actions/hosting-your-own-runners)
  - two micropython boards connected via USB 
  - created a [CI workflow](https://github.com/Josverl/pymakr-vsc/blob/next-gha/.github/workflows/hw_unit.yml) 

Initial results:  [![Node.js CI](https://github.com/Josverl/pymakr-vsc/actions/workflows/hw_unit.yml/badge.svg)](https://github.com/Josverl/pymakr-vsc/actions/workflows/hw_unit.yml)

Similar could be run on linux x64 , windows , and mac 

additional work is needed to log the test results in a standard format (JUnitXML seems the common supported format) so that other GHA tools can work with it to provide better feedback and to join additional platforms.

Also currently the action-runner is only installed for my fork of the repo, but it can also be installed for an organization (PyCom)

**Comment by Josverl** [comment]
@jakobrosenberg ,
note that this also may help simplify validation of version dependencies.
It appears that some unit tests are failing on node 16,x which might be relevant as VSCode 1.66.2 is also based on node 16.
```
Version: 1.66.2 (user setup)
Electron: 17.2.0
Node.js: 16.13.0

```
<img width="653" alt="image" src="https://user-images.githubusercontent.com/981654/164249256-58d8f1bc-3840-4c37-95ea-3566104d830b.png">

**Comment by jakobrosenberg** [comment]
This is great!! I'll fix the Node 16 issue.

**Comment by jakobrosenberg** [comment]
Do you know if it's possible to connect a self hosted runner to Github? If so we could replicate your setup and have full integration tests for PRs.

**Reply by Josverl** [reply]
> Do you know if it's possible to connect a self hosted runner to Github? If so we could replicate your setup and have full integration tests for PRs.

Yes , that is what I have done, 
I found it really much less complex than I expected, and it does not take much resources either.

I think it would make sense to run a few different GHA-SHRs with different setups and devices connected to them.
If you set this up we could work togetherby hooking up my runner to the pycom repo ( willl need to pass the setupkey securely ) 

The GH documentation is here : 
- https://docs.github.com/en/actions/hosting-your-own-runners 

I have a small writeup of my setup that I will check-in to the`next-gha`  branch as well.
- https://github.com/Josverl/pymakr-vsc/blob/next-gha/GHA-runner.md

**Comment by jakobrosenberg** [comment]
I have solved the Node 16 issues locally. You can try the branch here https://github.com/pycom/pymakr-vsc/tree/node-16

If it works for you, I'll merge it into the next-staging branch.

**Reply by Josverl** [reply]
## Unit test on Node16 
Unit tests work across all 8 environments 
node 14,15 X Mac , Win , Linux X64 Linux arm64 

https://github.com/Josverl/pymakr-vsc/actions/runs/2215712462
Note that I this is based on manual checks of the captured logs as the Success / Failure of the unit tests does not seem to be reported as an exit code.
is that something that you can add to probs ? 
Exit = 0 = All OK 
Exie != 0 = one or more errors 
(and Same for the integration tests ) 

the integration test in that branch does not seem to work at all 

Also not if I try to run it in the debugger on Winx64 , Linuxx64 or LinuxArm64 

## Integration tests on Node16 
**In the debugger** 
I think there may be an issue in `runIntegrationTests.js` not being referenced in the debug/unattended tests 
worked a better with : 
```
				"--extensionTestsPath=${workspaceFolder}/test/suite/runIntegrationTests.js"
```
then the tests do run in the debugger.

**CLI** 
running the `npm run test:integration` works 
Note it still reports errors across the different platforms and MCUs)
but that is good as that helps track down and solve these errors 



**CI Integration**
there are some issues with setting up and running the tests in CI , as there is a workaround needed for ARM64 in downloading and starting the correct .vscode-test binary for the platform.
Integration Log: https://gist.github.com/Josverl/ee539553fc31ad097d6d687aa1a3980d

I have not yet properly merged the different fixes for that with the node-16 one.
That is still merge-work to do ( or we could skip the Linux arm64 test for now)

**Reply by jakobrosenberg** [reply]
Great news! What kind of errors do you get with the integration test?

I'm currently rewriting the implementation. I'm hoping to get watch mode and interactive test filtering done by tomorrow. Can't make any promises though. Still a lot of gotchas when running the test runner inside the VSCode thread.

**Reply by Josverl** [reply]
note: I have a hunch that the failed test resolve mostly around :
- the setup of the filesystem ( root = \ vs root = \flash ) 
- MCU communication specifics ( ie send the reponse to `p[rint("foo") ` in a single write versus multiple writes 
possibly 
- MCU rest 
- the use of MCU specific micopython calls ?

it would be very good to have the test results in a structure (JUnit) test results document to find what tests are failing when

**Reply by Josverl** [reply]
> What kind of errors do you get with the integration test?
Integration Log: https://gist.github.com/Josverl/ee539553fc31ad097d6d687aa1a3980d

**Reply by Josverl** [reply]
Is there a way to flash a pycom firmware to a non-pycom esp32 device ?
as I do not have any pycom devices that may be one of the causes of inconcistency

**Reply by Josverl** [reply]
I found a generic pycom firmware to test with a pycom stock filesystem 
https://github.com/LIHUAWRITER/pycom-esp32-universal 
(I could not find firmwares to use in testing on the pycom site , is there something similar ?) 

Much better, only the `print("foo")` test fails 

**Tests passed: 24 / 25
Tests failed: 1 / 25**

https://gist.github.com/Josverl/82ed4e7fae41b39d1ea1c4ee38c5351c

**Comment by jakobrosenberg** [comment]
> Much better, only the print("foo") test fails

So close now. Any idea what could be the cause of this? What happens if you `print("foo")` manually in the terminal?

**Reply by Josverl** [reply]
I think the problem may be that: 
- the MCU send the response in multiple chunks , and not in a single string.
  - chunck 1 : `p`
  - chunck 2 : `rint("foo")\r\n`
  - chunck 3 : `'foo`
  - chunck 4 : `'\r\n>>>`
- that the function that is called in the test `nextTerminalData(device)` only recieves the first chunck or even stops after a single char 

![print_foo_chuncks](https://user-images.githubusercontent.com/981654/165394312-7e2cebcb-7321-47b3-be36-d8190e49bd5e.gif)

**Integration test on Winx64:** 
![print_foo](https://user-images.githubusercontent.com/981654/165392482-fdfaf765-0357-404f-ba1c-d920ff48a4f5.gif)

```
 ❌ can use print command   13 ms
node_modules/probs/lib/reporters/consoleReporter/consoleReporter.js:12
AssertionError: 'p' == 'print("foo")\r\nfoo\r\n>>> ' AssertionError [ERR_ASSERTION]: 'p' == 'print("foo")\r\nfoo\r\n>>> '
	at TestInstance.callback (c:\develop\vscode\pymakr-vsc\test\suite\integration\devices.test.js:44:16)
	at async TestInstance.runTestCallback (file:///c:/develop/vscode/pymakr-vsc/node_modules/probs/lib/Framework/TestRunner/TestInstance.js:76:13)
	at async TestInstance.run (file:///c:/develop/vscode/pymakr-vsc/node_modules/probs/lib/Framework/TestRunner/TestInstance.js:120:9)
	at async TestInstance.runChildren (file:///c:/develop/vscode/pymakr-vsc/node_modules/probs/lib/Framework/TestRunner/TestInstance.js:108:17)
	at async TestInstance.run (file:///c:/develop/vscode/pymakr-vsc/node_modules/probs/lib/Framework/TestRunner/TestInstance.js:121:9)
	at async TestInstance.runChildren (file:///c:/develop/vscode/pymakr-vsc/node_modules/probs/lib/Framework/TestRunner/TestInstance.js:108:17)
	at async TestInstance.run (file:///c:/develop/vscode/pymakr-vsc/node_modules/probs/lib/Framework/TestRunner/TestInstance.js:121:9)
	at async TestInstance.runChildren (file:///c:/develop/vscode/pymakr-vsc/node_modules/probs/lib/Framework/TestRunner/TestInstance.js:108:17)
	at async TestInstance.run (file:///c:/develop/vscode/pymakr-vsc/node_modules/probs/lib/Framework/TestRunner/TestInstance.js:121:9)
	at async TestInstance.runChildren (file:///c:/develop/vscode/pymakr-vsc/node_modules/probs/lib/Framework/TestRunner/TestInstance.js:108:17)
	at async TestInstance.run (file:///c:/develop/vscode/pymakr-vsc/node_modules/probs/lib/Framework/TestRunner/TestInstance.js:121:9)
	at async TestFile.run (file:///c:/develop/vscode/pymakr-vsc/node_modules/probs/lib/Framework/TestRunner/TestFile.js:28:9)
	at async QueueItem.run (file:///c:/develop/vscode/pymakr-vsc/node_modules/probs/lib/utils/QueueManager.js:8:36)
	at async QueueManager.checkQueue (file:///c:/develop/vscode/pymakr-vsc/node_modules/probs/lib/utils/QueueManager.js:41:13)
node_modules/probs/lib/reporters/consoleReporter/consoleReporter.js:14
          ✅ can disconnect    2 ms
no
```

**Comment by Josverl** [comment]
a quick update:
I gave Jest a try , first for the Unit tests.
## Unit Tests
The Unit tests were _very_ simple to update to run under Jest, with the follow pro's ::
- Test explorer 
- run tests one-by-one / watch / only failed etc 
- reports in multiple formats, Junit, github action 
- code coverage report 
- raises exitcode=1 so GHA can detect failures 

**Integration tests** 
The integration tests are more a bit more work to integrate, mostly as there is no formal API for Jest 

But there are a few templates in typescript that implement this, that need to be reworked into .js 
this is work in progress

###  test explorer
![image](https://user-images.githubusercontent.com/981654/165766978-d65873af-428c-4635-b462-c9b5b438e2fb.png)

### reports in Github Actions 
this simplifies spotting failing tests 
- there two tests are failing only on Mac and Node 14 or 16 
-see the results of this run: https://github.com/Josverl/pymakr-vsc/actions/runs/2236176529

![image](https://user-images.githubusercontent.com/981654/165767722-30913838-27b0-4a40-a6cc-d91829c6b23c.png)

**Comment by jakobrosenberg** [comment]
> the MCU send the response in multiple chunks , and not in a single string.
chunck 1 : p
chunck 2 : rint("foo")\r\n
chunck 3 : 'foo
chunck 4 : '\r\n>>>

I came across this too. It complicated detection of whether a device is running since the chunks can range from single characters, single lines and multiple lines. I think the affected test has been updated, so it might not affect you any more. If it does let me know and I'll have a look.

As for Jest, I started migrating my projects away from Jest last year because of scalability, performance and ESM issues. It's actually why I have started the probs project. Except for code coverage, probs has the same pros, though some are still in development. The lack of exitcode=1 was accidentally removed, but has been readded. I'll try get a release out asap. Apologies for the lack of documentation. This is on my todo list. Enabling watch mode (`--watch`) will provide a UI that lets you filter and watch tests.

**Reply by Josverl** [reply]
I believe you that jest ain't perfect , but after switching from probs to jest: 
- I could run individual tests 
- found and fixed dependencies between these tests (no docs that probs also could do that ) 
- I was able to find a few new issues.
- create and use mocks

As there are other extension authors that also use jest , which allows for some re-use.
I don't really care which testing framework is used , as long as
P1: 
- it can test, 
- mock, spy 

- there is some level documentation ( even if its : most is like jest / jasmine / but not foo and bar ) 
P2:
- report in a standard format ( to get structured reports via a tool such as [testspace](https://demo.testspace.com/)
- preferably with IDE integration in VSCode 

P3:
- code coverage
- there is some IP to reuse ( I can borrow from others working projects) 

I know that jest won't solve all issues either, for example: Currently working on mocking sufficient of vscode to allow for unit testing of the devices and devices store in order to find and resolve the issues on windows.

Not sure what the way forward would be on this. Its not for me to decide, but currently the lack of mocking / spying / stubbing holds me back from contributing.

One way could be that probs and jest can both be used for unit testing with a naming convention & configuration to seperate)
Once probs has the needed/used capabilities functionality, I suppose it should be as easy to convert from jest back to probs as it was for the reverse. 
To me that seems to be the most effective way to collaborate on PyMakr v-next
